test/dtypes.c

```diff
@@ -3819,6 +3821,193 @@ test_compound_18(void)
     return 1;
 } /* end test_compound_18() */
 
+/*-------------------------------------------------------------------------
+ * Function:    test_compound_19
+ *
+ * Purpose:     Tests that the library fails correctly when trying to
+ *              convert between compounds where none of the members in the
+ *              source compound appear in the destination compound.
+ *
+ * Return:      Success:        0
+ *              Failure:        number of errors
+ *
+ *-------------------------------------------------------------------------
+ */
+static int
+test_compound_19(void)
+{
+#if H5_SIZEOF_INT32_T != 0
+    H5T_order_t byte_order;
+    hsize_t     dset_dims[]  = { COMPOUND19_NUMELEMS };
+    hid_t       fid          = H5I_INVALID_HID;
+    hid_t       did          = H5I_INVALID_HID;
+    hid_t       space_id     = H5I_INVALID_HID;
+    hid_t       dst_compound = H5I_INVALID_HID;
+    herr_t      status;
+    char        filename[512];
+#if H5_SIZEOF_INT16_T != 0
+    hid_t compound_i16 = H5I_INVALID_HID;
+#endif
+#if H5_SIZEOF_INT32_T != 0
+    hid_t compound_i32 = H5I_INVALID_HID;
+#endif
+#if H5_SIZEOF_INT64_T != 0
+    hid_t compound_i64 = H5I_INVALID_HID;
```

Comment:
Fixed size types are used here just to make sure that the optimizations in struct conversions around smaller or larger destination types don't potentially break this in the future